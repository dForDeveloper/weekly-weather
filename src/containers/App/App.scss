.App--header {
  height: 80px;
  box-shadow: $heavy-shadow;
  background-color: $blue;
}

.h1 {
  font-size: 32px;
  margin: 0;
  padding: 24px 0 24px 24px;
  color: #fff;
  font-weight: 300;
}

.App--h2 {
  margin-top: 16px;
  text-align: center;
}

.App--div--darksky {
  text-align: center;
  margin: 64px 0 32px;

  a {
    color: $blue;
  }
}