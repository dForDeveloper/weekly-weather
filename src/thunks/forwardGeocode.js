import { fetchData } from '../utils/api';
import { setCoordinates, setCity, setError } from '../actions';

export const forwardGeocode = (query) => {
  return async (dispatch) => {
    query = query.replace(/\s/g, '+');
    query = query.replace(/,/g, '%2C');
    try {
      const url = `http://localhost:3001/api/v1/forwardgeocode/${query}`;
      const geocodeData = await fetchData(url);
      const { lat: latitude, lng: longitude } = geocodeData[0].geometry;
      const { city, state_code } = geocodeData[0].components;
      dispatch(setCity(`${city}, ${state_code}`));
      dispatch(setCoordinates({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}