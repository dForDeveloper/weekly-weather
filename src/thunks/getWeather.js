import { fetchData } from '../utils/api';
import { cleanData } from '../utils/dataCleaner';
import { setWeather, setError, toggleLoading } from '../actions';

export const getWeather = ({ latitude: lat, longitude: lon }) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const url = `http://localhost:3001/api/v1/weather/${lat}/${lon}`;
      const weather = await fetchData(url);
      dispatch(setWeather(cleanData(weather)));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message))      
    }
    dispatch(toggleLoading(false));
  }
}