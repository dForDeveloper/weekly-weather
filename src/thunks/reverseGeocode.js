import { fetchData } from '../utils/api';
import { setCity, setError } from '../actions';

export const reverseGeocode = ({ latitude: lat, longitude: lon }) => {
  return async (dispatch) => {
    try {
      const url = `http://weekly-weather.herokuapp.com/api/v1/reversegeocode/${lat}/${lon}`;
      const geocodeData = await fetchData(url);
      const { city, state_code } = geocodeData[0].components;
      dispatch(setCity(`${city}, ${state_code}`));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}