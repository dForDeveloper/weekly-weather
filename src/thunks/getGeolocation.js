import { setCoordinates, setError, toggleLoading } from '../actions';
import { reverseGeocode } from './reverseGeocode';
import { getWeather } from './getWeather';

export const getGeolocation = ({ latitude, longitude }, history) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try{
      dispatch(setCoordinates({ latitude, longitude }));
      await dispatch(reverseGeocode({ latitude, longitude }, history));
      await dispatch(getWeather({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(toggleLoading(false));
  }
}