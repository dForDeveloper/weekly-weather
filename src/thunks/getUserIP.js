import { fetchData } from '../utils/api';
import { setCoordinates, setError, toggleLoading } from '../actions';
import { reverseGeocode } from '../thunks/reverseGeocode';
import { getWeather } from '../thunks/getWeather';

export const getUserIP = (history) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try{
      const url = 'https://weekly-weather.herokuapp.com/api/v1/ip';
      const { latitude, longitude } = await fetchData(url);
      dispatch(setCoordinates({ latitude, longitude }));
      await dispatch(reverseGeocode({ latitude, longitude }, history));
      await dispatch(getWeather({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(toggleLoading(false));
  }
}