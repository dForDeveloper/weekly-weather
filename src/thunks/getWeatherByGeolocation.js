import { setCoordinates, setError, toggleLoading } from '../actions';
import { reverseGeocode } from '../thunks/reverseGeocode';
import { getWeather } from '../thunks/getWeather';

export const getWeatherByGeolocation = ({ latitude, longitude }) => {
  return (dispatch) => {
    dispatch(toggleLoading(true));
    try{
      dispatch(setCoordinates({ latitude, longitude }));
      dispatch(reverseGeocode({ latitude, longitude }));
      dispatch(getWeather({ latitude, longitude }));
      dispatch(setError(''));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(toggleLoading(false));
  }
}