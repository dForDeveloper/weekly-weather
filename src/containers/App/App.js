import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter, Redirect } from 'react-router-dom';
import Search from '../Search/Search';
import WeatherContainer from '../../components/WeatherContainer/WeatherContainer';
import { getUserIP } from '../../thunks/getUserIP';
import { getWeatherByGeolocation } from '../../thunks/getWeatherByGeolocation';
import PropTypes from 'prop-types';

export class App extends Component {
  constructor() {
    super();
    this.state = {
      locationGranted: true
    }
  }

  componentDidMount() {
    this.getGeolocation();
  }

  getGeolocation = () => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(async (position) => {
        const { latitude, longitude } = await position.coords;
        this.props.getWeatherByGeolocation({ latitude, longitude });
      },
        () => {
          this.setState({ locationGranted: false }, () => this.props.getUserIP())
        }
      );
    } else {
      this.props.getUserIP();
    }
  }

  render() {
    const { userLocation, weather, location, error, isLoading } = this.props;
    const { city } = userLocation;
    const redirectPath = city.replace(/\s/g, '');
    const shouldRedirect = !location.pathname.includes(redirectPath) && city;
    return (
      <div className="App">
        <header className="App--header">
          <h1 className="h1">Weekly Weather</h1>
        </header>
        <Search />
        {!isLoading && 
          <div>
            {shouldRedirect && <Redirect to={redirectPath} />}
            {weather.today && !error &&
              <WeatherContainer city={city} weather={weather} />}
            {!this.state.locationGranted && !userLocation.city && 
              <h2 className="App--h2">
                Share your location for automatic redirection to your city
              </h2>
            }
            {error && <h2 className="App--h2">No results found</h2>}
          </div>}
        {isLoading && <h2 className="App--h2">Loading...</h2>}
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  userLocation: state.location,
  weather: state.weather,
  error: state.error,
  isLoading: state.isLoading
});

export const mapDispatchToProps = (dispatch) => ({
  getUserIP: () => dispatch(getUserIP()),
  getWeatherByGeolocation: ({ latitude, longitude }) => dispatch(getWeatherByGeolocation({ latitude, longitude }))
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));

App.propTypes = {
  userLocation: PropTypes.object,
  weather: PropTypes.object,
  error: PropTypes.string,
  isLoading: PropTypes.bool,
  getUserIP: PropTypes.func,
  getWeatherByGeolocation: PropTypes.func,
}