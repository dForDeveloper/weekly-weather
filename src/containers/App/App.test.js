import React from 'react';
import ReactDOM from 'react-dom';
import { App } from './App';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { createStore } from 'redux';
import rootReducer from '../../reducers';
import { getUserIP } from '../../thunks/getUserIP';

jest.mock('../../thunks/getUserIP.js');

describe('App', () => {
  const mockProps = {
    userLocation: { 
      latitude: 0,
      longitude: 0,
      city: 'Denver, CO'
    },
    weather: {
      currently: { temperature: 39 }
    },
    error: '',
    getUserIP: jest.fn(),
    location: { pathname: '/' }
  }

  it('renders without crashing', () => {
    const div = document.createElement('div');
    const store = createStore(rootReducer);
    const provider = (
      <Provider store={store}>
        <BrowserRouter>
          <App {...mockProps} />
        </BrowserRouter>
      </Provider>
    );
    ReactDOM.render(provider , div);
    ReactDOM.unmountComponentAtNode(div);
  });
});