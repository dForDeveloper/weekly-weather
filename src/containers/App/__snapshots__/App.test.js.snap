// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot when isLoading is false 1`] = `
<div
  className="App"
>
  <header
    className="App--header"
  >
    <h1
      className="h1"
    >
      Weekly Weather
    </h1>
  </header>
  <Connect(Search) />
  <div>
    <WeatherContainer
      city="denver"
      weather={
        Object {
          "graphData": Array [
            Object {
              "data": Array [
                Object {
                  "x": "7:00 PM",
                  "y": 37,
                },
                Object {
                  "x": "8:00 PM",
                  "y": 33,
                },
                Object {
                  "x": "9:00 PM",
                  "y": 31,
                },
                Object {
                  "x": "10:00 PM",
                  "y": 29,
                },
              ],
              "id": "",
            },
          ],
          "maxTemp": 42,
          "minTemp": 19,
          "today": Object {
            "day": "Sat",
            "feelsLike": 21,
            "high": 37,
            "humidity": "66%",
            "icon": "snow",
            "low": 19,
            "precipProbability": "51%",
            "precipType": "snow",
            "summary": "Partly Cloudy",
            "sunrise": "6:56 AM",
            "sunset": "5:17 PM",
            "temperature": 31,
            "wind": "3 mph WNW",
          },
          "week": Array [
            Object {
              "day": "Sun",
              "high": 42,
              "humidity": "51%",
              "icon": "partly-cloudy-day",
              "low": 23,
              "precipProbability": "4%",
              "precipType": "snow",
              "summary": "Partly cloudy until afternoon.",
              "sunrise": "6:31 AM",
              "sunset": "5:24 PM",
              "wind": "5 mph SSE",
            },
          ],
        }
      }
    />
  </div>
</div>
`;

exports[`App should match the snapshot when isLoading is true 1`] = `
<div
  className="App"
>
  <header
    className="App--header"
  >
    <h1
      className="h1"
    >
      Weekly Weather
    </h1>
  </header>
  <Connect(Search) />
  <h1>
    Loading...
  </h1>
</div>
`;

exports[`App should match the snapshot when there is an error 1`] = `
<div
  className="App"
>
  <header
    className="App--header"
  >
    <h1
      className="h1"
    >
      Weekly Weather
    </h1>
  </header>
  <Connect(Search) />
  <div>
    <h1>
      No results found
    </h1>
  </div>
</div>
`;

exports[`App should redirect when pathname does not match the city 1`] = `
<div
  className="App"
>
  <header
    className="App--header"
  >
    <h1
      className="h1"
    >
      Weekly Weather
    </h1>
  </header>
  <Connect(Search) />
  <div>
    <Redirect
      push={false}
      to="denver"
    />
    <WeatherContainer
      city="denver"
      weather={
        Object {
          "graphData": Array [
            Object {
              "data": Array [
                Object {
                  "x": "7:00 PM",
                  "y": 37,
                },
                Object {
                  "x": "8:00 PM",
                  "y": 33,
                },
                Object {
                  "x": "9:00 PM",
                  "y": 31,
                },
                Object {
                  "x": "10:00 PM",
                  "y": 29,
                },
              ],
              "id": "",
            },
          ],
          "maxTemp": 42,
          "minTemp": 19,
          "today": Object {
            "day": "Sat",
            "feelsLike": 21,
            "high": 37,
            "humidity": "66%",
            "icon": "snow",
            "low": 19,
            "precipProbability": "51%",
            "precipType": "snow",
            "summary": "Partly Cloudy",
            "sunrise": "6:56 AM",
            "sunset": "5:17 PM",
            "temperature": 31,
            "wind": "3 mph WNW",
          },
          "week": Array [
            Object {
              "day": "Sun",
              "high": 42,
              "humidity": "51%",
              "icon": "partly-cloudy-day",
              "low": 23,
              "precipProbability": "4%",
              "precipType": "snow",
              "summary": "Partly cloudy until afternoon.",
              "sunrise": "6:31 AM",
              "sunset": "5:24 PM",
              "wind": "5 mph SSE",
            },
          ],
        }
      }
    />
  </div>
</div>
`;
